### PR TITLE
Issue90 type information

### PR DESCRIPTION
**Pull request by KasperJanssens**
Cary,

First try of the type information. It seems quite stable, been using it for a few days (provided the configuration is correct, I suppose, didn't test what happens when ghc-modi is not correctly configured).

I bound the type info call to the DocumentationProvider as well as to an action (alt - equals, like scala). I prefer the action, I think the documentation provider doesn't work so well. There are tests of the parsing of the output of ghc-modi, but not really of the documentationprovider, basically because of the abundance of static calls and the fact that I think they can only be mocked while testing, and statics can only be mocked through Powermock if I recall correctly, which would mean an extra test dependency and so on and so forth, so I left that to be your call.

Also, there is a weird behaviour that getting the editor creates a stack trace of around 5 kilometers long, something that seems like a threading issue, but the function seems to work. I don't really know why the stack trace happens, all the more because it only happens when calling the type information through the documentation provider (same code path is used when the action is called, but no stack trace). Maybe you know more what could go wrong, it looks like something intellij-related.

So, consider this a "request for comment" more than a pull request ;-)

Kasper

### CODE REVIEW

File: META-INF/plugin.xml [hunk 1]
```diff
@@ -113,6 +113,10 @@
             text="Restart ghc-modi" description="Restarts a disabled ghc-modi process">
       <add-to-group group-id="ToolsMenu" anchor="last"/>
     </action>
+      <action id="TypeInfo" class="com.haskforce.highlighting.annotation.external.TypeInfoAction" text="Type Info">
+          <add-to-group group-id="ViewMenu" anchor="first"/>
+          <keyboard-shortcut first-keystroke="alt EQUALS" keymap="$default"/>
+      </action>
```
Comment (carymrobbins) [inline META-INF/plugin.xml]: Could you update the registration keybindings to include `F11` for Mac?  The `⌥ =` combination doesn't work well on Mac without tampering with the global keyboard configuration.  Also (nit-pick), could you indent the XML at the same level as the other configurations?
```xml
<action id="TypeInfo" class="com.haskforce.highlighting.annotation.external.TypeInfoAction" text="Type Info">
  <add-to-group group-id="ViewMenu" anchor="first"/>
  <keyboard-shortcut first-keystroke="F11" keymap="Mac OS X"/>
  <keyboard-shortcut first-keystroke="F11" keymap="Mac OS X 10.5+"/>
  <keyboard-shortcut first-keystroke="alt EQUALS" keymap="$default"/>
</action>
```

File: src/com/haskforce/highlighting/annotation/external/TypeInfoAction.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import com.intellij.openapi.actionSystem.AnAction;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.fileEditor.FileEditorManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.popup.JBPopup;
+import com.intellij.openapi.ui.popup.JBPopupFactory;
+import com.intellij.ui.awt.RelativePoint;
+
+public class TypeInfoAction extends AnAction {
```
Comment (carymrobbins) [inline src/com/haskforce/highlighting/annotation/external/TypeInfoAction.java]: Could we put this class under `com.haskforce.actions` instead?

File: tests/com/haskforce/highlighting/annotation/external/GhcUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+import com.intellij.openapi.editor.LogicalPosition;
+import com.intellij.openapi.editor.VisualPosition;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * Created by kasper on 12/24/14.
+ */
+public class GhcUtilTest {
```
Comment (carymrobbins) [inline tests/com/haskforce/highlighting/annotation/external/GhcUtilTest.java]: Nice!  Thanks for adding tests.  Be sure to add this class to **tests/com/haskforce/HaskellTestCase.java**

File: src/com/haskforce/highlighting/annotation/HaskellDocumentationProvider.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import org.jetbrains.annotations.Nullable;
+
+import java.awt.*;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+public class HaskellDocumentationProvider extends AbstractDocumentationProvider {
```
Comment (carymrobbins) [inline src/com/haskforce/highlighting/annotation/HaskellDocumentationProvider.java]: This doesn't seem to work for me...not sure if I'm doing something wrong.

**Comment by carymrobbins** [comment]
Nice work @KasperJanssens!  Sorry, barely getting around to trying this out.

I've updated the **Files Changed** tab with comments and questions.  Overall, this is really good stuff.  I'm glad that you have tests, you've refactored some crufty stuff, and tapped into popups to provide a custom UI.

File: tests/com/haskforce/highlighting/annotation/external/GhcUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
+package com.haskforce.highlighting.annotation.external;
+
+import com.intellij.openapi.editor.LogicalPosition;
+import com.intellij.openapi.editor.VisualPosition;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * Created by kasper on 12/24/14.
```
Comment (pjonsson) [inline tests/com/haskforce/highlighting/annotation/external/GhcUtilTest.java]: I love JavaDoc. Can you drop your name/date (we have that in Git annotate) and add documentation that is more in line with the JavaDoc in other similar files?

File: src/com/haskforce/highlighting/annotation/HaskellDocumentationProvider.java (hunk 4 above)
Comment (KasperJanssens) [inline src/com/haskforce/highlighting/annotation/HaskellDocumentationProvider.java]: Can you tell me in a little more detail what doesn't seem to work? I'll try to look into it as soon as I have beaten my desktop back into submission again, it doesn't want to build nor run the plugin anymore. My laptop still has it and runs so likely a config issue, as soon as I've got it sorted out I'll check it.

Comment (KasperJanssens) [inline src/com/haskforce/highlighting/annotation/HaskellDocumentationProvider.java]: Never mind, reproduced, fixed. I performed a clean up that was a bit too harsh, and totally wrong. Anyways, should work now. Let me know.

**Comment by KasperJanssens** [comment]
Concerning the tests, I think I could a little better but I'm afraid that to test some things like those static calls to GhcMod(i) I need to depend on a mocking framework like mockito or powermock, don't know what you think about that. I also didn't realize the tests were JUnit 3, any reason for that, is that enforced by the intellij plugin stuff or so? I don't know whether it would be a problem for the newer mockito or powermock versions. Anyway, something to consider in case we would like to extend the tests with mocks.
Concerning the refactor, I just looked at my freshly added stuff, didn't check whether there was a need to refactor some of the older stuff. In case you see something or think of something let me know. 

I should have taken care of all review comments and fixed them, if anything is still left, let me know.

File: tests/com/haskforce/highlighting/annotation/external/GhcUtilTest.java (hunk 5 above)
Comment (KasperJanssens) [inline tests/com/haskforce/highlighting/annotation/external/GhcUtilTest.java]: I mainly added javadoc on class level, not on method level or so. If you think more is needed, let me know.

**Comment by pjonsson** [comment]
The test suite already uses mockups to some extent since there wouldn't be any IntelliJ-environment to interact with without a mockup. We use whatever that comes with IntelliJ since bringing in more recent versions and/or adding more dependencies risk version mismatches that are difficult to debug. Those kind of problems might not be a problem for regulars but definitely risk scaring away newcomers.

Don't be afraid to use one commit for each kind of change--that usually makes things easier to review and makes git bisect work better in a distant future.

**Comment by carymrobbins** [comment]
> Don't be afraid to use one commit for each kind of change

+1 on this.  We can always rebase to clean up before merging into master.

@KasperJanssens - I really don't like most mocking frameworks.  It's a sign of bad design, which is true of GhcMod(i).  We should be using proper dependency inversion.  I'm not sure what that looks like exactly just yet, but I'd rather focus on that than building tests around a bad design. /rant

**Comment by KasperJanssens** [comment]
Alright, in this case I already squashed my commits before handing them over to you guys, will refrain from that in the future. 
Agreed about the mocking, the alternative is indeed dependency injection and test doubles. Not sure neither how that will best be done with those static functions, but the bottom line is that for this pull request I will not start adding mock tests ;-).

**Comment by carymrobbins** [comment]
Sounds good, thanks @KasperJanssens.  Once we get this merged in I'll start working on refactoring the GhcMod classes.

**Pull request closed without merging**